### PR TITLE
Add entra-agent-id skill

### PR DESCRIPTION
**Pull request by ArLucaID**
Adds a new plugin skill: **entra-agent-id** — Microsoft Entra Agent ID setup, agent identity provisioning, and OAuth 2.0 token exchange via Microsoft Graph v1.0.

## Whats in this PR

- `plugin/skills/entra-agent-id/SKILL.md` — core skill doc
- `plugin/skills/entra-agent-id/references/`
  - `runtime-token-exchange.md` — two-step `fmi_path` exchange (autonomous + OBO, cross-tenant)
  - `oauth2-token-flow.md` — MI + WIF (production) and client secret (local dev)
  - `obo-blueprint-setup.md` — configuring the Blueprint as an OAuth2 API for OBO
  - `sdk-sidecar.md` — Microsoft Entra SDK for AgentID polyglot sidecar (`mcr.microsoft.com/entra-sdk/auth-sidecar`)
  - `known-limitations.md` — documented gaps
- `tests/entra-agent-id/` — unit, trigger, and integration test scaffold
- `tests/skills.json` — registers the skill and adds it to the daily integration schedule
- `.github/CODEOWNERS` — `@ArLucaID` and `@jongio`

## Validation

- `npm run checkCopilotCliCharBudget` -> 18,355 / 20,000
- `npm run typecheck` -> clean
- `npm run lint` -> clean
- `npx gulp` -> version stamped `skills/entra-agent-id 1.1.5`
- `npm run test:unit -- --testPathPatterns=entra-agent-id` -> 33 / 33 passing
- `npm run tokens -- check` -> within limits

## Notes

- Uses Microsoft Graph v1.0 typed endpoints (`/applications/microsoft.graph.agentIdentityBlueprint` etc.), not `/beta` with `@odata.type` discriminators.
- Matches the `entra-app-registration` skill structure and frontmatter conventions.

### CODE REVIEW

File: plugin/skills/entra-agent-id/version.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+  "version": "1.1",
```
Comment (tmeschter) [inline plugin/skills/entra-agent-id/version.json]: Since this is a new skill this can be "1.0".

File: tests/entra-agent-id/integration.test.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+if (skipTests && skipReason) {
+  console.log(`⏭️  Skipping integration tests: ${skipReason}`);
+}
+
+const describeIntegration = skipTests ? describe.skip : describe;
+
+describeIntegration(`${SKILL_NAME}_ - Integration Tests`, () => {
+  const agent = useAgentRunner();
+
```
Comment (kvenkatrajan) [inline tests/entra-agent-id/integration.test.ts:49]: Consider adding other integration tests like - 

create agent identity blueprint
create agent identity
testing runtime authentication if possible

**Comment by kvenkatrajan** [comment]
## Skill Review — PR #2033: Add entra-agent-id skill

### ✅ What Looks Good

1. **Excellent conceptual model** — The Blueprint → BlueprintPrincipal → Agent Identity hierarchy is clearly documented with ASCII art and a summary table
2. **Thorough troubleshooting** — Error table covers 8 specific errors with cause and fix, matching real-world failure modes (AADSTS codes)
3. **Strong routing boundary** — Clean `DO NOT USE FOR` clause disambiguates from `entra-app-registration`, and domain-specific keywords (`fmi_path`, `BlueprintPrincipal`, `agentIdentityBlueprint`) minimize routing conflicts
4. **Complete test scaffold** — Unit, trigger, and integration tests all present with edge cases (empty prompt, long prompt, case insensitivity)
5. **Progressive reference loading** — 5 reference files linked contextually (not bulk-loaded), following JIT disclosure pattern

---

### 🔴 Critical

**1. `sdk-sidecar.md` exceeds hard token limit**

**File(s):** `plugin/skills/entra-agent-id/references/sdk-sidecar.md`
**Issue:** ~2,892 tokens (hard limit: 2,000). This file covers SDK configuration, 6 code patterns in 3 languages, Docker Compose, Kubernetes deployment, security guidance, and troubleshooting — too much for a single reference file.
**Suggested Fix:** Split into two files:
- `references/sdk-sidecar.md` — Architecture, configuration, query params, endpoint reference (~1,200 tokens)
- `references/sdk-sidecar-deployment.md` — Code patterns, Docker Compose, Kubernetes YAML, security (~1,600 tokens)

Update SKILL.md links accordingly.

---

### 🟠 High

**2. Description exceeds 60-word guideline**

**File(s):** `plugin/skills/entra-agent-id/SKILL.md:3`
**Issue:** Description is 81 words (guideline: ≤60). Long descriptions consume extra tokens on every routing evaluation.
**Suggested Fix:** Trim to ~55 words by consolidating trigger phrases.

**3. Missing required SKILL.md sections for service skills**

**File(s):** `plugin/skills/entra-agent-id/SKILL.md`
**Issue:** Missing three required sections per the checklist:
- **Quick Reference** — No summary table (Best for, Primary capabilities, MCP tools)
- **When to Use This Skill** — No explicit activation scenarios list
- **MCP Tools** — No MCP tool table (should state explicitly that this skill guides manual Graph API calls, or reference `mcp_azure_mcp_documentation` for Graph lookups)

**Suggested Fix:** Add a Quick Reference table after the title, a "When to Use This Skill" bulleted list before "Before You Start", and either an MCP Tools table or a note explaining this skill guides manual Graph API calls rather than MCP tool invocations.

**4. `runtime-token-exchange.md` approaching hard token limit**

**File(s):** `plugin/skills/entra-agent-id/references/runtime-token-exchange.md`
**Issue:** ~1,712 tokens (hard limit: 2,000). Close to the ceiling with limited room for future additions.
**Suggested Fix:** Consider extracting the cross-tenant section (~30 lines) into a separate `references/cross-tenant.md` if the file grows further. No action needed now, but flag for awareness.

---

### 🟡 Medium

**5. Three reference files exceed soft token limits**

**File(s):** `known-limitations.md` (~1,186), `oauth2-token-flow.md` (~1,156), `obo-blueprint-setup.md` (~1,083)
**Issue:** All slightly above the 1,000-token soft limit. Not blocking, but watch for growth.

**6. Uses `USE FOR:` instead of preferred `WHEN:` trigger format**

**File(s):** `plugin/skills/entra-agent-id/SKILL.md:3`
**Issue:** Checklist prefers `WHEN:` with quoted phrases. However, `USE FOR:` is used by the existing `entra-app-registration` skill, so this is consistent within the Entra skill family. Low priority.

---

### 🟢 Low

**7. SKILL.md references `mcp_azure_mcp_documentation` tool without formal MCP Tools section**

**File(s):** `plugin/skills/entra-agent-id/SKILL.md:34`
**Issue:** The tool is referenced in "Before You Start" but not documented in a dedicated MCP Tools section.

---

### Code Review (automated)

- ✅ No hardcoded credentials — all samples use env vars/placeholders
- ✅ All 5 reference files properly linked from SKILL.md
- ✅ CODEOWNERS alphabetically sorted
- ✅ skills.json correctly registers in both list and integration schedule
- ✅ Graph API endpoints and OAuth2 flows are correct
- ✅ Test patterns match established repo conventions

### Summary

| Severity | Count |
|----------|-------|
| 🔴 Critical | 1 |
| 🟠 High | 3 |
| 🟡 Medium | 2 |
| 🟢 Low | 1 |

**Overall Assessment: Request Changes** — The `sdk-sidecar.md` token budget violation and missing required SKILL.md sections need to be addressed before merge. The content quality and routing boundaries are excellent.

**Comment by ArLucaID** [comment]
@kvenkatrajan: Thank you for the review of the skill, I've gone ahead and applied all the requested changes. :)

File: plugin/skills/entra-agent-id/SKILL.md [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+
+Create and manage OAuth 2.0-capable identities for AI agents using Microsoft Graph. Every agent instance gets a distinct identity, audit trail, and independently-scoped permission grants.
+
+## Before You Start
+
+Search the `microsoft-docs` MCP for current Agent ID documentation:
+- "Microsoft Entra Agent ID setup instructions"
+- "Microsoft Entra SDK for AgentID"
+
+Verify request bodies and endpoints against the installed SDK version — Graph API shapes evolve.
```
[the remark this answers is not in the document]
Comment (ArLucaID) [reply plugin/skills/entra-agent-id/SKILL.md:51]: Addressed in commit ebe0922 — updated to mcp_azure_mcp_documentation (the repo's documented Azure docs MCP tool name). Thanks!

File: plugin/skills/entra-agent-id/SKILL.md [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+```
+
+## Core Workflow
+
+### Step 1: Create Agent Identity Blueprint
+
+Use the typed endpoint. Sponsors must be **Users** at Blueprint creation.
+
+```python
+import subprocess
```
[the remark this answers is not in the document]
Comment (ArLucaID) [reply plugin/skills/entra-agent-id/SKILL.md:133]: Addressed in commit ebe0922 — added import requests and import subprocess to the Step 1 snippet so it's copy-pastable independently. Thanks!

File: plugin/skills/entra-agent-id/version.json [hunk 5]
```diff
@@ -0,0 +1,6 @@
...
+  "pathFilters": [
+    "."
+  ]
+}
```
[the remark this answers is not in the document]
Comment (ArLucaID) [reply plugin/skills/entra-agent-id/version.json:6]: Addressed in commit 0fbeff7 — set baseline to 1.0 since this is a new skill. NBGV stamps the build-time version on top of this baseline. The PR description's earlier mention of 1.1.5 was the auto-stamped value from the older 1.1 baseline; with 1.0 baseline it stamps as 1.0.x.

File: tests/entra-agent-id/integration.test.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+const skipTests = shouldSkipIntegrationTests();
+const skipReason = getIntegrationSkipReason();
+
+if (skipTests && skipReason) {
+  console.log(`⏭️  Skipping integration tests: ${skipReason}`);
+}
+
+const describeIntegration = skipTests ? describe.skip : describe;
+
+describeIntegration(`${SKILL_NAME}_ - Integration Tests`, () => {
```
[the remark this answers is not in the document]
Comment (ArLucaID) [reply tests/entra-agent-id/integration.test.ts:47]: Tried this in commit 0fbeff7 but had to revert — the repo's custom/integration-test-name ESLint rule requires the suffix pattern ^[a-z0-9-]+_[a-z0-9-]* - Integration Tests$, i.e. the trailing _ is mandatory. Removing it produces: Top-level describe name 'entra-agent-id - Integration Tests' does not match the required pattern. Sibling skills (e.g. irunway-aks-setup) follow the same ${SKILL_NAME}_ - Integration Tests convention. Happy to change if you'd like to update the lint rule.

File: tests/entra-agent-id/integration.test.ts (hunk 2 above)
Comment (ArLucaID) [reply tests/entra-agent-id/integration.test.ts:49]: Addressed in commit 8e72a3c — added 5 new integration test prompts:

- Blueprint creation with sponsor + BlueprintPrincipal (specific Step 1 prompt)
- Per-instance Agent Identity creation (specific Step 3 prompt)
- OBO runtime authentication (delegated path)
- Per-Agent-Identity permission grant (`User.Read.All` via `appRoleAssignments`)
- `AADSTS82001` troubleshooting (error-driven invocation)

Each runs 5x with invocation-rate threshold 0.8, same pattern as the existing tests. Total integration prompts is now 8.

File: plugin/skills/entra-agent-id/SKILL.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (ArLucaID) [reply plugin/skills/entra-agent-id/SKILL.md:133]: Addressed in commit ebe0922 — added import requests and import subprocess to the Step 1 snippet so it's copy-pastable independently. Thanks!

File: plugin/skills/entra-agent-id/version.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (ArLucaID) [reply plugin/skills/entra-agent-id/version.json:6]: Addressed in commit 0fbeff7 — set baseline to 1.0 since this is a new skill. NBGV stamps the build-time version on top of this baseline. The PR description's earlier mention of 1.1.5 was the auto-stamped value from the older 1.1 baseline; with 1.0 baseline it stamps as 1.0.x.

File: tests/entra-agent-id/integration.test.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (ArLucaID) [reply tests/entra-agent-id/integration.test.ts:47]: Tried this in commit 0fbeff7 but had to revert — the repo's custom/integration-test-name ESLint rule requires the suffix pattern ^[a-z0-9-]+_[a-z0-9-]* - Integration Tests$, i.e. the trailing _ is mandatory. Removing it produces: Top-level describe name 'entra-agent-id - Integration Tests' does not match the required pattern. Sibling skills (e.g. irunway-aks-setup) follow the same ${SKILL_NAME}_ - Integration Tests convention. Happy to change if you'd like to update the lint rule.

File: plugin/skills/entra-agent-id/SKILL.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (ArLucaID) [reply plugin/skills/entra-agent-id/SKILL.md:133]: Addressed in commit ebe0922 — added import requests and import subprocess to the Step 1 snippet so it's copy-pastable independently. Thanks!

File: plugin/skills/entra-agent-id/version.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (ArLucaID) [reply plugin/skills/entra-agent-id/version.json:6]: Addressed in commit 0fbeff7 — set baseline to 1.0 since this is a new skill. NBGV stamps the build-time version on top of this baseline. The PR description's earlier mention of 1.1.5 was the auto-stamped value from the older 1.1 baseline; with 1.0 baseline it stamps as 1.0.x.

[the remark this answers is not in the document]
Comment (ArLucaID) [reply plugin/skills/entra-agent-id/version.json:6]: Addressed in commit 0fbeff7 — set baseline to 1.0 since this is a new skill. NBGV stamps the build-time version on top of this baseline. The PR description's earlier mention of 1.1.5 was the auto-stamped value from the older 1.1 baseline; with 1.0 baseline it stamps as 1.0.x.

[the remark this answers is not in the document]
Comment (ArLucaID) [reply plugin/skills/entra-agent-id/version.json:6]: Addressed in commit 0fbeff7 — set baseline to 1.0 since this is a new skill. NBGV stamps the build-time version on top of this baseline. The PR description's earlier mention of 1.1.5 was the auto-stamped value from the older 1.1 baseline; with 1.0 baseline it stamps as 1.0.x.

File: tests/entra-agent-id/integration.test.ts (hunk 2 above)
Comment (ArLucaID) [reply tests/entra-agent-id/integration.test.ts:49]: Followed up in commit 1f44d77 with content-assertion tests modeled on the azure-deploy pattern you pointed me at. Each test runs a single prompt (no early-termination so we get the full reply) and asserts on the agent's response text:

- Blueprint creation prompt → response must contain `microsoft.graph.agentIdentityBlueprint`, `microsoft.graph.agentIdentityBlueprintPrincipal`, and `sponsors@odata.bind`
- Per-instance Agent Identity prompt → response must contain `microsoft.graph.agentIdentity` and `agentIdentityBlueprintId`
- `fmi_path` exchange prompt → response must contain `fmi_path` + `client_credentials` + `/.default`/`AzureADTokenExchange`
- Auth guidance prompt → response must steer toward `client_credentials` / `Connect-MgGraph` (the paths that work today)
- Permission grant prompt → response must use `appRoleAssignments` or `oauth2PermissionGrants`

Helpers live in tests/entra-agent-id/utils.ts following the same shape as tests/azure-deploy/utils.ts (regex patterns + getAllAssistantMessages).

Note: stronger live-Graph artifact verification (actually call Graph post-run to confirm a Blueprint exists) is blocked today by the `Directory.AccessAsUser.All` rejection on Agent ID CRUD via `az login`/OIDC — that block is being removed in the coming weeks. Will add a live-Graph variant in a follow-up PR once it lands.

**Comment by kvenkatrajan** [comment]
As discussed in chat with @ArLucaID  the end to end test will be added after a fix has been rolled out on 5/5

**Review by kvenkatrajan — APPROVED** [review]
